pmlc/dialect/tile/transforms/padding.cc

```diff
@@ -0,0 +1,213 @@
+  BoundRange(int64_t min, int64_t max) : min(min), max(max) {}
+  BoundRange operator+(const BoundRange &rhs) const {
+    return {min + rhs.min, max + rhs.max};
+  }
+  BoundRange operator*(const BoundRange &rhs) const {
+    return {std::min(std::min(min * rhs.min, max * rhs.min),
+                     std::min(min * rhs.max, max * rhs.max)),
+            std::max(std::max(min * rhs.min, max * rhs.min),
+                     std::max(min * rhs.max, max * rhs.max))};
+  }
+};
+
+class ComputeRangeVisitor
+    : public mlir::AffineExprVisitor<ComputeRangeVisitor, BoundRange> {
+public:
+  explicit ComputeRangeVisitor(std::vector<BoundRange> dimRanges)
+      : dimRanges(std::move(dimRanges)) {}
+
+  BoundRange visitMulExpr(mlir::AffineBinaryOpExpr expr) {
+    return visit(expr.getLHS()) * visit(expr.getRHS());
+  }
+  BoundRange visitAddExpr(mlir::AffineBinaryOpExpr expr) {
+    return visit(expr.getLHS()) + visit(expr.getRHS());
+  }
+  BoundRange visitDimExpr(mlir::AffineDimExpr expr) {
+    return dimRanges[expr.getPosition()];
+  }
+  BoundRange visitSymbolExpr(mlir::AffineSymbolExpr expr) {
+    llvm_unreachable("Invalid symbol in ComputeRangeVisitor");
+  }
+  BoundRange visitConstantExpr(mlir::AffineConstantExpr expr) {
+    return BoundRange(expr.getValue());
+  }
+  BoundRange visitModExpr(mlir::AffineBinaryOpExpr expr) {
+    llvm_unreachable("Invalid mod in ComputeRangeVisitor");
+  }
+  BoundRange visitCeilDivExpr(mlir::AffineBinaryOpExpr expr) {
+    llvm_unreachable("Invalid ceil in ComputeRangeVisitor");
+  }
+  BoundRange visitFloorDivExpr(mlir::AffineBinaryOpExpr expr) {
+    llvm_unreachable("Invalid floor in ComputeRangeVisitor");
+  }
+
+private:
+  std::vector<BoundRange> dimRanges;
+};
+
+llvm::SmallVector<BoundRange, 4>
+computeBounds(AffineMap access, AffineMap lower, AffineMap upper) {
+  // We only handle the 'easy' cases.
+  assert(access.getNumSymbols() == 0);
+  assert(lower.getNumSymbols() == 0);
+  assert(upper.getNumSymbols() == 0);
+  assert(lower.getNumDims() == 0);
+  assert(upper.getNumDims() == 0);
+  size_t idxs = access.getNumDims();
+  size_t size = access.getNumResults();
+  assert(lower.getNumResults() == idxs);
+  assert(upper.getNumResults() == idxs);
+  assert(access.isPureAffine());
+  std::vector<BoundRange> dimRanges;
+  for (size_t i = 0; i < idxs; i++) {
+    int64_t lowerBoundInt =
+        lower.getResult(i).cast<mlir::AffineConstantExpr>().getValue();
+    int64_t upperBoundInt =
+        upper.getResult(i).cast<mlir::AffineConstantExpr>().getValue();
+    dimRanges.emplace_back(lowerBoundInt, upperBoundInt);
+  }
+  ComputeRangeVisitor visitor(std::move(dimRanges));
+  llvm::SmallVector<BoundRange, 4> outRanges;
+  for (size_t i = 0; i < size; i++) {
+    outRanges.push_back(visitor.visit(access.getResult(i)));
+  }
+  return outRanges;
+}
+
+struct PaddingInfo {
+  llvm::SmallVector<int64_t, 4> padBelow;
+  llvm::SmallVector<int64_t, 4> padAbove;
+};
```

Comment:
Shall we put agg kinds inside PaddingInfo to avoid DenseMap<AggregationKind, PaddingInfo>? It could be  a simple "uint32_t aggKinds", which is a bitmap for all agg kinds. Then we can check if the aggKinds has multiple 1s, i.e., check if it is power of 2 using  a simple trick aggKinds^(aggKinds-1) == 0. If aggKinds has multiple 1s, padBelow and padAbove are useless and can be overwritten.

Reply:
I considered something along these lines, however I decided that it was premature, since we may someday soon want to pad two different ways, with one of them being a copy, and one being inline for example.  In order to do that, we would need to track more, so I just tracked it.